Initial implementation of d2l-autocomplete and d2l-autocomplete-text-…

AdmiralCoco opened the pull request:
…input

dlockhart commented on d2l-autocomplete-text-input.html:
```diff
@@ -0,0 +1,90 @@
+			id="d2l-autocomplete-text-input"
+			role="combobox"
+			aria-expanded$="[[suggestionsVisible]]"
+			aria-owns="d2l-autocomplete"
```
For `aria-owns`, `aria-controls` and the `for` on `<d2l-autocomplete>`, if more than one of these ends up on the same page the IDs aren't going to be unique and you'll get collisions. I'd recommend using our [polymer behaviour[(https://github.com/Brightspace/d2l-polymer-behaviors-ui/blob/master/d2l-id.html) to generate a unique ID for the autocomplete and text input.

But even then... I'm not sure this is going to work correctly with shadow DOM enabled. Screen readers are going to expect the `for` to point at a particular HTML input element, and when they see a `d2l-text-input` I don't think that'll work. This is a known limitation of shadow DOM + a lot of the ARIA stuff. IDs simply don't work across shadow boundaries.

dlockhart replied:
If your testing ends up showing that it works, that's amazing! Just make sure shadow DOM is actually enabled (it's off by default in Polymer 1, on by default in Polymer 2) and that the browser you're testing with supports it (Chrome and Safari only currently, Firefox very soon).

dlockhart commented on d2l-autocomplete.html:
```diff
@@ -1,13 +1,70 @@
+				position: relative;
+			}
+
+			.bolded {
```
A couple things here:

1. Since when shadow DOM is disabled these classes end up in the global namespace, we try to always make them unique to avoid someone else putting a similar class on the page and having it impact our elements (or vice versa).
2. Try to name CSS classes based on their purpose, not the style that you want to apply. The style may change in the future, and then the class name doesn't make sense anymore.

Recommendation: change it to something like `d2l-autocomplete-suggestion-highlighted`

dlockhart commented on d2l-autocomplete.html:
```diff
@@ -1,13 +1,70 @@
+			.d2l-autocomplete-list {
+				-moz-box-sizing: border-box;
+				-ms-box-sizing: border-box;
+				-webkit-box-sizing: border-box;
```
You shouldn't need prefixes for box sizing:
https://caniuse.com/#feat=css3-boxsizing

dlockhart commented on d2l-autocomplete.html:
```diff
@@ -1,13 +1,70 @@
+				white-space: nowrap;
+			}
+
+			.d2l-autocomplete-suggestion.selected,
```
Do you need the CSS `selected` class at all, could you just use an attribute selector instead? e.g. `d2l-autocomplete-suggestion[aria-selected="true"]`

dlockhart commented on d2l-autocomplete.html:
```diff
@@ -21,16 +78,324 @@ <h2>Hello [[prop1]]!</h2>
+			connectedCallback() {
+				super.connectedCallback();
+				Polymer.RenderStatus.afterNextRender(this, () => {
+					this._input = this.parentNode.querySelector(`#${this.for}`);
```
I have a feeling this is going to be really problematic with shadow DOM...

dlockhart commented on d2l-autocomplete.html:
```diff
@@ -21,16 +78,324 @@ <h2>Hello [[prop1]]!</h2>
+			}
+
+			_onInput(event) {
+				const handler = filter => () => (this._filter = filter);
```
Arrow functions likely won't fly in IE11:
https://caniuse.com/#feat=arrow-functions

dbatiste commented on d2l-autocomplete.html:
```diff
@@ -1,13 +1,63 @@
+				position: relative;
+			}
+
+			.d2l-autocomplete-list {
```
Will the list always show below the input?  If so, what happens in the scenario where autocomplete is rendered at the bottom of the page?  i.e. does it overflow the viewport?

An alternative option is to use the reusable [d2l-dropdown-content](https://github.com/BrightspaceUI/dropdown/blob/master/d2l-dropdown-content.html) or [D2L.PolymerBehaviors.DropdownContentBehavior](https://github.com/BrightspaceUI/dropdown/blob/master/d2l-dropdown-content-behavior.html) that contains the positioning logic.

AdmiralCoco replied:
Good point - I'll see if we can use the `DropdownContentBehavior` here

dbatiste replied:
Cool.  There is also the [D2L.PolymerBehaviors.DropdownOpenerBehavior](https://github.com/BrightspaceUI/dropdown/blob/master/d2l-dropdown-opener-behavior.html) that can be used for the opener, which I _think_ is the text input in this case.  The openers tend to be pretty simple as a result.

I think My Courses did something very similar in the All Course view.  Eventually, it may be updated to use your new component.

svanherk commented:
Hey, just came across this PR... heads up, the `d2l-text-input` component is archived and was just removed from BSI.  The replacement is `d2l-input-text` from https://github.com/BrightspaceUI/inputs

AdmiralCoco commented:
I've updated the component to basically be a `d2l-dropdown` wrapper on an input element - it now looks something like:
```html
<d2l-autocomplete>
    <input slot="input">
</d2l-autocomplete>
``` 

The autocomplete adds the `d2l-dropdown-opener` class to the `slot="input"` element, which allows it to utilize the `d2l-dropdown-opener-behaviour`

I've also updated to use the `d2l-input-text` and renamed some files accordingly

(Also apologies for the slow update, I've been working on another project)
